### DISCUSSION TITLE
Mock Resources

### DISCUSSION THREAD

**Discussion by psamuels00** [mockoon/mockoon#844] [Ideas]
Hi.  Thank you for putting this tool together and out on open source.  I looked at many options, but few as polished, both GUI and CLI.  The rules system is great, especially the ability to reference the request number as a target.  The wind went right out of my sails, though, when I realized it is scoped to the endpoint.

Let's say we have a GET request in a REST API that returns the status of a resource.  If the server is expected to receive requests for a single resource only, then the request number can be used to control the sequence of responses representing a well-defined pattern of status values.  If, however, we want the server to handle multiple resources at once, with requests for various resources arriving in any order, then the request number is suddenly of little use.

It would be nice if multiple resources could be handled at once, each generating the same pattern of responses.  In fact, it would really be nice if they didn't have to generate the _same_ pattern, but rather, _some_ well-defined pattern of responses.

To make this possible, I've implemented [a tweak in a fork](https://github.com/psamuels00/mockoon-w-mock-resources/tree/feature/000_add_mock_resources) that, essentially, adds recognition of "mock resources" and scopes the request number to the mock resource instead of to the endpoint.  See [MOCK_RESOURCES.md](https://github.com/psamuels00/mockoon-w-mock-resources/blob/feature/000_add_mock_resources/MOCK_RESOURCES.md) for details and the tiniest bit of rambling commentary.  See the [diffs here](https://github.com/mockoon/mockoon/compare/main...psamuels00:mockoon-w-mock-resources:feature/000_add_mock_resources).

I may be reading too deeply, but I believe this issue relates to discussions #458, #601, and #819, as well as issues #124, #158, and #783.

In many cases, mock resources can be used in place of CRUD, data buckets, and/or session-scoped environment variables.  At least that's my hypothesis.  As always, there are pros and cons.

On the plus side, the POC is implemented with no change at all to the GUI.  For a complete solution, some change would be needed, but this could be as simple as a single Boolean flag in the settings to indicate how the request number should be interpreted.

A second point is more serious.  A stateless server scales very well.  I admit I'm not aware of how the data buckets are implemented, or how the CRUD and environment variables will be implemented, but what I do know is if you let people store stuff, they will store stuff.  My implementation of mock resources requires server-side storage as well, but at least the impact is minimized since only an ID and request number is saved for each resource.

On the negative side, the setup for mock resources can become complicated, depending on what you are trying to achieve.  See the bottom of [MOCK_RESOURCES.md](https://github.com/psamuels00/mockoon-w-mock-resources/blob/feature/000_add_mock_resources/MOCK_RESOURCES.md) for examples.

There are many other pros and cons to this approach.  What do you think?  I'm curious to know if this topic has surfaced before?

Thanks!

**Comment by 255kb** [comment]
Sorry for the delay in my reply.
As you saw, this topic has surfaced before but is scattered among multiple issues (CRUD endpoints, etc.). What users are asking is sometimes vague, but one thing is certain, CRUD endpoints are the most requested feature by far (though, it is usually expressed not by an issue upvote but rather by asking why a PUT/POST is not persisted and is available in the next GET). The usual expectation is something similar to what json-server does, which seems to be the "standard". At least it's widely used and has been implemented in other mocking tools.
This is what is planned in the roadmap. The data buckets will make the implementation easier by providing templates that persist between calls, which was not the case until now. 
I have to be honest. After reading your explanation, I am not sure I understood everything. It seems your implementation is deeply linked to the request number, and not necessarily only to CRUD operations. Also, the configuration of mock resources (end of the file) seems a bit cryptic, at least for me. 

Mockoon key point being simplicity (though it's not always easy to keep things simple...), I think it's better to implement CRUD endpoints that require no configuration and are behaving in line with the leading tools on the market. 

However, we could definitely add new rules, like server scoped request number, and the possibility to reset these counters, maybe as part of #629.
